{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "import requests\n",
    "import shutil\n",
    "import datetime as dt\n",
    "import sys\n",
    "import numpy as np\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_PROJECTS = [\n",
    "    'TCGA-ACC', 'TCGA-BLCA', 'TCGA-BRCA', 'TCGA-CESC', 'TCGA-CHOL', \n",
    "    'TCGA-COAD', 'TCGA-DLBC', 'TCGA-ESCA', 'TCGA-GBM', 'TCGA-HNSC', \n",
    "    'TCGA-KICH', 'TCGA-KIRC', 'TCGA-KIRP', 'TCGA-LAML', 'TCGA-LGG', \n",
    "    'TCGA-LIHC', 'TCGA-LUAD', 'TCGA-LUSC', 'TCGA-MESO', 'TCGA-OV', \n",
    "    'TCGA-PAAD', 'TCGA-PCPG', 'TCGA-PRAD', 'TCGA-READ', 'TCGA-SARC', \n",
    "    'TCGA-SKCM', 'TCGA-STAD', 'TCGA-TGCT', 'TCGA-THCA', 'TCGA-THYM', \n",
    "    'TCGA-UCEC', 'TCGA-UCS', 'TCGA-UVM'\n",
    "]\n",
    "OMICS = {\n",
    "    'methylation450': 'DNA_Methylation_450',\n",
    "    'methylation27': 'DNA_Methylation_27',\n",
    "    'star_fpkm': 'Gene_Expression',\n",
    "    'clinical': 'Clinical_Data',\n",
    "    'mirna': 'miRNA_Data'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_tcga_data(project, data_type):\n",
    "    directory = OMICS[data_type]\n",
    "    tcga_download_endpoint = f'https://gdc-hub.s3.us-east-1.amazonaws.com/download/{project}.{data_type}.tsv.gz'\n",
    "    output_file_path = f'UCSC_TCGA_Data/{directory}/{project}.gz'\n",
    "\n",
    "    decompressed_file_path = output_file_path.replace('.gz', '')\n",
    "    if os.path.exists(decompressed_file_path):\n",
    "        print(f\"File already exists: {decompressed_file_path}\")\n",
    "        return\n",
    "    \n",
    "    response = requests.get(tcga_download_endpoint, stream=True)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "        print(f\"Starting the download for project {project}\")\n",
    "        os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
    "        with open(output_file_path, 'wb') as f:\n",
    "            for chunk in response.iter_content(chunk_size=8192):\n",
    "                f.write(chunk)\n",
    "        \n",
    "        print(f\"Downloaded file saved to {output_file_path}\")\n",
    "        \n",
    "        with gzip.open(output_file_path, 'rb') as f_in:\n",
    "            with open(output_file_path.replace('.gz', ''), 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "        \n",
    "        print(f\"Decompressed file saved to {output_file_path.replace('.gz', '')}\")\n",
    "\n",
    "        os.remove(output_file_path)\n",
    "        print(f\"Removed compressed file: {output_file_path}\")\n",
    "    else:\n",
    "        print(f\"Failed to download file. Status code: {response.status_code}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting the download for project TCGA-ACC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-ACC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-ACC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-ACC.gz\n",
      "Starting the download for project TCGA-ACC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-ACC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-ACC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-ACC.gz\n",
      "Starting the download for project TCGA-ACC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-ACC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-ACC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-ACC.gz\n",
      "Starting the download for project TCGA-BLCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-BLCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-BLCA\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-BLCA.gz\n",
      "Starting the download for project TCGA-BLCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-BLCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-BLCA\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-BLCA.gz\n",
      "Starting the download for project TCGA-BLCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-BLCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-BLCA\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-BLCA.gz\n",
      "Starting the download for project TCGA-BRCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-BRCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-BRCA\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-BRCA.gz\n",
      "Starting the download for project TCGA-BRCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-BRCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-BRCA\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-BRCA.gz\n",
      "Starting the download for project TCGA-BRCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-BRCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-BRCA\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-BRCA.gz\n",
      "Starting the download for project TCGA-CESC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-CESC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-CESC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-CESC.gz\n",
      "Starting the download for project TCGA-CESC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-CESC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-CESC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-CESC.gz\n",
      "Starting the download for project TCGA-CESC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-CESC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-CESC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-CESC.gz\n",
      "Starting the download for project TCGA-CHOL\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-CHOL.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-CHOL\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-CHOL.gz\n",
      "Starting the download for project TCGA-CHOL\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-CHOL.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-CHOL\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-CHOL.gz\n",
      "Starting the download for project TCGA-CHOL\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-CHOL.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-CHOL\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-CHOL.gz\n",
      "Starting the download for project TCGA-COAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-COAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-COAD\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-COAD.gz\n",
      "Starting the download for project TCGA-COAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-COAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-COAD\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-COAD.gz\n",
      "Starting the download for project TCGA-COAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-COAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-COAD\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-COAD.gz\n",
      "Starting the download for project TCGA-DLBC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-DLBC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-DLBC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-DLBC.gz\n",
      "Starting the download for project TCGA-DLBC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-DLBC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-DLBC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-DLBC.gz\n",
      "Starting the download for project TCGA-DLBC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-DLBC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-DLBC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-DLBC.gz\n",
      "Starting the download for project TCGA-ESCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-ESCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-ESCA\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-ESCA.gz\n",
      "Starting the download for project TCGA-ESCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-ESCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-ESCA\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-ESCA.gz\n",
      "Starting the download for project TCGA-ESCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-ESCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-ESCA\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-ESCA.gz\n",
      "Starting the download for project TCGA-GBM\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-GBM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-GBM\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-GBM.gz\n",
      "Starting the download for project TCGA-GBM\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-GBM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-GBM\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-GBM.gz\n",
      "Starting the download for project TCGA-GBM\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-GBM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-GBM\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-GBM.gz\n",
      "Starting the download for project TCGA-HNSC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-HNSC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-HNSC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-HNSC.gz\n",
      "Starting the download for project TCGA-HNSC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-HNSC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-HNSC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-HNSC.gz\n",
      "Starting the download for project TCGA-HNSC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-HNSC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-HNSC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-HNSC.gz\n",
      "Starting the download for project TCGA-KICH\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KICH.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KICH\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KICH.gz\n",
      "Starting the download for project TCGA-KICH\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-KICH.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-KICH\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-KICH.gz\n",
      "Starting the download for project TCGA-KICH\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-KICH.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-KICH\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-KICH.gz\n",
      "Starting the download for project TCGA-KIRC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KIRC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KIRC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KIRC.gz\n",
      "Starting the download for project TCGA-KIRC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-KIRC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-KIRC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-KIRC.gz\n",
      "Starting the download for project TCGA-KIRC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-KIRC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-KIRC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-KIRC.gz\n",
      "Starting the download for project TCGA-KIRP\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KIRP.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KIRP\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-KIRP.gz\n",
      "Starting the download for project TCGA-KIRP\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-KIRP.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-KIRP\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-KIRP.gz\n",
      "Starting the download for project TCGA-KIRP\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-KIRP.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-KIRP\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-KIRP.gz\n",
      "Starting the download for project TCGA-LAML\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LAML.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LAML\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LAML.gz\n",
      "Starting the download for project TCGA-LAML\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LAML.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LAML\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-LAML.gz\n",
      "Starting the download for project TCGA-LAML\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LAML.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LAML\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-LAML.gz\n",
      "Starting the download for project TCGA-LGG\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LGG.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LGG\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LGG.gz\n",
      "Starting the download for project TCGA-LGG\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LGG.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LGG\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-LGG.gz\n",
      "Starting the download for project TCGA-LGG\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LGG.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LGG\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-LGG.gz\n",
      "Starting the download for project TCGA-LIHC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LIHC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LIHC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LIHC.gz\n",
      "Starting the download for project TCGA-LIHC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LIHC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LIHC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-LIHC.gz\n",
      "Starting the download for project TCGA-LIHC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LIHC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LIHC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-LIHC.gz\n",
      "Starting the download for project TCGA-LUAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LUAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LUAD\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LUAD.gz\n",
      "Starting the download for project TCGA-LUAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LUAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LUAD\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-LUAD.gz\n",
      "Starting the download for project TCGA-LUAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LUAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LUAD\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-LUAD.gz\n",
      "Starting the download for project TCGA-LUSC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LUSC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LUSC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-LUSC.gz\n",
      "Starting the download for project TCGA-LUSC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LUSC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-LUSC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-LUSC.gz\n",
      "Starting the download for project TCGA-LUSC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LUSC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-LUSC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-LUSC.gz\n",
      "Starting the download for project TCGA-MESO\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-MESO.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-MESO\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-MESO.gz\n",
      "Starting the download for project TCGA-MESO\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-MESO.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-MESO\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-MESO.gz\n",
      "Starting the download for project TCGA-MESO\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-MESO.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-MESO\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-MESO.gz\n",
      "Starting the download for project TCGA-OV\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-OV.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-OV\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-OV.gz\n",
      "Starting the download for project TCGA-OV\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-OV.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-OV\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-OV.gz\n",
      "Starting the download for project TCGA-OV\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-OV.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-OV\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-OV.gz\n",
      "Starting the download for project TCGA-PAAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PAAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PAAD\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PAAD.gz\n",
      "Starting the download for project TCGA-PAAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-PAAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-PAAD\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-PAAD.gz\n",
      "Starting the download for project TCGA-PAAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-PAAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-PAAD\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-PAAD.gz\n",
      "Starting the download for project TCGA-PCPG\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PCPG.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PCPG\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PCPG.gz\n",
      "Starting the download for project TCGA-PCPG\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-PCPG.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-PCPG\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-PCPG.gz\n",
      "Starting the download for project TCGA-PCPG\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-PCPG.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-PCPG\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-PCPG.gz\n",
      "Starting the download for project TCGA-PRAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PRAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PRAD\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-PRAD.gz\n",
      "Starting the download for project TCGA-PRAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-PRAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-PRAD\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-PRAD.gz\n",
      "Starting the download for project TCGA-PRAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-PRAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-PRAD\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-PRAD.gz\n",
      "Starting the download for project TCGA-READ\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-READ.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-READ\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-READ.gz\n",
      "Starting the download for project TCGA-READ\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-READ.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-READ\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-READ.gz\n",
      "Starting the download for project TCGA-READ\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-READ.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-READ\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-READ.gz\n",
      "Starting the download for project TCGA-SARC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-SARC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-SARC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-SARC.gz\n",
      "Starting the download for project TCGA-SARC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-SARC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-SARC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-SARC.gz\n",
      "Starting the download for project TCGA-SARC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-SARC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-SARC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-SARC.gz\n",
      "Starting the download for project TCGA-SKCM\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-SKCM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-SKCM\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-SKCM.gz\n",
      "Starting the download for project TCGA-SKCM\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-SKCM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-SKCM\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-SKCM.gz\n",
      "Starting the download for project TCGA-SKCM\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-SKCM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-SKCM\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-SKCM.gz\n",
      "Starting the download for project TCGA-STAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-STAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-STAD\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-STAD.gz\n",
      "Starting the download for project TCGA-STAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-STAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-STAD\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-STAD.gz\n",
      "Starting the download for project TCGA-STAD\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-STAD.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-STAD\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-STAD.gz\n",
      "Starting the download for project TCGA-TGCT\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-TGCT.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-TGCT\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-TGCT.gz\n",
      "Starting the download for project TCGA-TGCT\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-TGCT.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-TGCT\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-TGCT.gz\n",
      "Starting the download for project TCGA-TGCT\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-TGCT.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-TGCT\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-TGCT.gz\n",
      "Starting the download for project TCGA-THCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-THCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-THCA\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-THCA.gz\n",
      "Starting the download for project TCGA-THCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-THCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-THCA\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-THCA.gz\n",
      "Starting the download for project TCGA-THCA\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-THCA.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-THCA\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-THCA.gz\n",
      "Starting the download for project TCGA-THYM\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-THYM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-THYM\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-THYM.gz\n",
      "Starting the download for project TCGA-THYM\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-THYM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-THYM\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-THYM.gz\n",
      "Starting the download for project TCGA-THYM\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-THYM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-THYM\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-THYM.gz\n",
      "Starting the download for project TCGA-UCEC\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UCEC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UCEC\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UCEC.gz\n",
      "Starting the download for project TCGA-UCEC\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-UCEC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-UCEC\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-UCEC.gz\n",
      "Starting the download for project TCGA-UCEC\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-UCEC.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-UCEC\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-UCEC.gz\n",
      "Starting the download for project TCGA-UCS\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UCS.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UCS\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UCS.gz\n",
      "Starting the download for project TCGA-UCS\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-UCS.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-UCS\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-UCS.gz\n",
      "Starting the download for project TCGA-UCS\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-UCS.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-UCS\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-UCS.gz\n",
      "Starting the download for project TCGA-UVM\n",
      "Downloaded file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UVM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UVM\n",
      "Removed compressed file: UCSC_TCGA_Data/DNA_Methylation_450/TCGA-UVM.gz\n",
      "Starting the download for project TCGA-UVM\n",
      "Downloaded file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-UVM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/Gene_Expression/TCGA-UVM\n",
      "Removed compressed file: UCSC_TCGA_Data/Gene_Expression/TCGA-UVM.gz\n",
      "Starting the download for project TCGA-UVM\n",
      "Downloaded file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-UVM.gz\n",
      "Decompressed file saved to UCSC_TCGA_Data/miRNA_Data/TCGA-UVM\n",
      "Removed compressed file: UCSC_TCGA_Data/miRNA_Data/TCGA-UVM.gz\n"
     ]
    }
   ],
   "source": [
    "for project in TCGA_PROJECTS:\n",
    "    download_tcga_data(project, \"methylation450\")\n",
    "    # download_tcga_data(project, \"methylation27\")\n",
    "    download_tcga_data(project, \"star_fpkm\")\n",
    "    download_tcga_data(project, \"mirna\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Join the projects for DNA Methylation 27k and 450k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_illumina27_patients(project: str):\n",
    "    file_path = Path('UCSC_TCGA_Data').joinpath('DNA_Methylation_27', project)\n",
    "    if not file_path.exists():\n",
    "        return []\n",
    "    else:\n",
    "        df = pl.read_csv(file_path, separator='\\t')\n",
    "        return df.columns\n",
    "\n",
    "def join_methylation_projects(project):\n",
    "    output_path = os.path.join('UCSC_TCGA_Data', 'DNA_Methylation')\n",
    "    output_file = Path(output_path).joinpath(project)\n",
    "\n",
    "    # check whether the output path exists\n",
    "    if not Path(output_path).exists():\n",
    "        Path(output_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "    # read and join the files\n",
    "    file_path_27 = os.path.join('UCSC_TCGA_Data', 'DNA_Methylation_27', project)\n",
    "    file_path_450 = os.path.join('UCSC_TCGA_Data', 'DNA_Methylation_450', project)\n",
    "\n",
    "    # validate if the joining procedure shall be started\n",
    "    if not (Path(file_path_27).exists() or Path(file_path_450).exists()):\n",
    "        print(f'Both files do not exists to join. They might have been joined already or accidently been removed')\n",
    "        return\n",
    "\n",
    "    # check whether a 27k file could be downloaded\n",
    "    if not Path(file_path_27).exists():\n",
    "        print(f'No Illumina 27k file was available for project {project}, moving Illumina 450k experiment to {output_file} instead')\n",
    "        Path(file_path_450).replace(output_file)\n",
    "    else:\n",
    "        print(f'Reading and joining both DNA Methylation files for project {project}')\n",
    "        df_27 = pl.read_csv(file_path_27, separator='\\t')\n",
    "        df_450 = pl.read_csv(file_path_450, separator='\\t')\n",
    "    \n",
    "        df = df_27.join(df_450, on='Composite Element REF', how='full', coalesce=True)\n",
    "        print(f'Files joined from shapes {df_27.shape} and {df_450.shape} to shape {df.shape}')\n",
    "        \n",
    "        df.write_csv(output_file, separator='\\t')\n",
    "        print(f'Methylation files joined into {output_file}')\n",
    "    \n",
    "def remove_separate_dna_methylation_projects(project):\n",
    "    file_path = Path('UCSC_TCGA_Data')\n",
    "    print(f'Removing Illumina 27k and Illumina 450k files of {project}')\n",
    "    file_path.joinpath('DNA_Methylation_27', project).unlink(missing_ok=True)\n",
    "    file_path.joinpath('DNA_Methylation_450', project).unlink(missing_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-ACC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-BLCA\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-BRCA\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-CESC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-CHOL\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-COAD\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-DLBC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-ESCA\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-GBM\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-HNSC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-KICH\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-KIRC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-KIRP\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-LAML\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-LGG\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-LIHC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-LUAD\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-LUSC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-MESO\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-OV\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-PAAD\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-PCPG\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-PRAD\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-READ\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-SARC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-SKCM\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-STAD\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-TGCT\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-THCA\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-THYM\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-UCEC\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-UCS\n",
      "Both files do not exists to join. They might have been joined already or accidently been removed\n",
      "Removing Illumina 27k and Illumina 450k files of TCGA-UVM\n"
     ]
    }
   ],
   "source": [
    "illumina_27_patients = []\n",
    "for project in TCGA_PROJECTS:\n",
    "    illumina_27_patients.extend(get_illumina27_patients(project))\n",
    "    join_methylation_projects(project)\n",
    "    remove_separate_dna_methylation_projects(project)\n",
    "illumina_27_patients = [item for item in illumina_27_patients if item != 'Composite Element REF']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process the Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNA Methylation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine which probes to keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_probes_to_remove_set():\n",
    "    hg38_genes_file_path = 'humanmethylation450_15017482_v1-2.csv'\n",
    "    df = pl.read_csv(hg38_genes_file_path, separator=',', null_values=['NA'], skip_lines=7, columns=['IlmnID', 'CHR', 'UCSC_RefGene_Name'])\n",
    "\n",
    "    no_gene_associated = set()\n",
    "    sexual_chr_probes = set()\n",
    "    chrM_probes = set()\n",
    "    non_cpg_probes = set()\n",
    "    non_standard_transcript_probes = set()\n",
    "\n",
    "    # standard_transcript_types = {\"protein_coding\", \"lncRNA\", \"miRNA\"}\n",
    "\n",
    "    for row in df.iter_rows(named=True):\n",
    "        chromosome = row['CHR']  \n",
    "        probe_id = str(row['IlmnID']) \n",
    "        gene_id = str(row['UCSC_RefGene_Name'])\n",
    "        gene_names = gene_id.split(\";\") \n",
    "        # transcript_types = row['transcriptTypes']\n",
    "            \n",
    "        if gene_id == 'None':\n",
    "            no_gene_associated.add(probe_id)\n",
    "            continue\n",
    "\n",
    "        if not probe_id.startswith('cg'):\n",
    "            non_cpg_probes.add(probe_id)\n",
    "            continue\n",
    "\n",
    "        if chromosome == 'chrY' or chromosome=='chrX':\n",
    "            sexual_chr_probes.add(probe_id)\n",
    "            continue\n",
    "\n",
    "        if chromosome=='chrM':\n",
    "            chrM_probes.add(probe_id)\n",
    "            continue\n",
    "\n",
    "        if len(gene_names) == 1: # Remove AL/AF lncRNA\n",
    "            if '.' in gene_names[0]:\n",
    "                non_standard_transcript_probes.add(probe_id)\n",
    "\n",
    "    print(f\"Number of probes without gene: {len(no_gene_associated)}\")\n",
    "    print(f\"Number of non standard probes: {len(non_cpg_probes)}\")\n",
    "    print(f\"Number of sexual chromosome probes: {len(sexual_chr_probes)}\")\n",
    "    print(f\"Number of M chromosome probes: {len(chrM_probes)}\")\n",
    "    print(f\"Number of non standard transcript type probes: {len(non_standard_transcript_probes)}\")\n",
    "\n",
    "    probes_to_remove = no_gene_associated.union(non_cpg_probes).union(sexual_chr_probes). \\\n",
    "        union(chrM_probes).union(non_standard_transcript_probes)\n",
    "    probes_to_keep = {*df.select('IlmnID').filter(~df['IlmnID'].is_in(probes_to_remove)).to_series().to_list()}\n",
    "\n",
    "    return probes_to_remove, probes_to_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of probes without gene: 120568\n",
      "Number of non standard probes: 1217\n",
      "Number of sexual chromosome probes: 0\n",
      "Number of M chromosome probes: 0\n",
      "Number of non standard transcript type probes: 25\n"
     ]
    }
   ],
   "source": [
    "probes_to_remove, probes_to_keep = create_probes_to_remove_set()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove unwanted probes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_valid_probe(probe_name):\n",
    "    return probe_name not in probes_to_remove\n",
    "\n",
    "def process_project_methylation_files(protocol, project):\n",
    "    print(f'Processing data for project {project}')\n",
    "    project_file_path = f'UCSC_TCGA_Data/{protocol}/{project}'\n",
    "    output_file_path = (f'UCSC_TCGA_Data_Clean/{protocol}/'\n",
    "        f'{project}_filtered_methylation.tsv')\n",
    "    \n",
    "    if os.path.exists(output_file_path):\n",
    "        return\n",
    "\n",
    "    if not os.path.exists(project_file_path):\n",
    "        return\n",
    "\n",
    "    df = pl.read_csv(project_file_path, separator='\\t',\n",
    "                     null_values=['NA'])\n",
    "    print(f'Initial shape {df.shape}')\n",
    "    df = df.rename({'Composite Element REF': 'Probe'}) \n",
    "\n",
    "    df_filtered = df.filter(~df['Probe'].is_in(probes_to_remove))\n",
    "\n",
    "    os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
    "    print(df_filtered.shape)\n",
    "    df_filtered.write_csv(output_file_path, separator='\\t')\n",
    "    print(f'Finished writing data to file {output_file_path}')\n",
    "\n",
    "    # Remove the original file to save up space  \n",
    "    #os.remove(project_file_path)\n",
    "    #print(f\"Deleted original file {project_file_path} to free up space.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing data for project TCGA-ACC\n",
      "Processing data for project TCGA-ACC\n",
      "Processing data for project TCGA-ACC\n",
      "Processing data for project TCGA-BLCA\n",
      "Processing data for project TCGA-BLCA\n",
      "Processing data for project TCGA-BLCA\n",
      "Processing data for project TCGA-BRCA\n",
      "Processing data for project TCGA-BRCA\n",
      "Processing data for project TCGA-BRCA\n",
      "Processing data for project TCGA-CESC\n",
      "Processing data for project TCGA-CESC\n",
      "Processing data for project TCGA-CESC\n",
      "Processing data for project TCGA-CHOL\n",
      "Processing data for project TCGA-CHOL\n",
      "Processing data for project TCGA-CHOL\n",
      "Processing data for project TCGA-COAD\n",
      "Processing data for project TCGA-COAD\n",
      "Processing data for project TCGA-COAD\n",
      "Processing data for project TCGA-DLBC\n",
      "Processing data for project TCGA-DLBC\n",
      "Processing data for project TCGA-DLBC\n",
      "Processing data for project TCGA-ESCA\n",
      "Processing data for project TCGA-ESCA\n",
      "Processing data for project TCGA-ESCA\n",
      "Processing data for project TCGA-GBM\n",
      "Processing data for project TCGA-GBM\n",
      "Processing data for project TCGA-GBM\n",
      "Processing data for project TCGA-HNSC\n",
      "Processing data for project TCGA-HNSC\n",
      "Processing data for project TCGA-HNSC\n",
      "Processing data for project TCGA-KICH\n",
      "Processing data for project TCGA-KICH\n",
      "Processing data for project TCGA-KICH\n",
      "Processing data for project TCGA-KIRC\n",
      "Processing data for project TCGA-KIRC\n",
      "Processing data for project TCGA-KIRC\n",
      "Processing data for project TCGA-KIRP\n",
      "Processing data for project TCGA-KIRP\n",
      "Processing data for project TCGA-KIRP\n",
      "Processing data for project TCGA-LAML\n",
      "Processing data for project TCGA-LAML\n",
      "Processing data for project TCGA-LAML\n",
      "Processing data for project TCGA-LGG\n",
      "Processing data for project TCGA-LGG\n",
      "Processing data for project TCGA-LGG\n",
      "Processing data for project TCGA-LIHC\n",
      "Processing data for project TCGA-LIHC\n",
      "Processing data for project TCGA-LIHC\n",
      "Processing data for project TCGA-LUAD\n",
      "Processing data for project TCGA-LUAD\n",
      "Processing data for project TCGA-LUAD\n",
      "Processing data for project TCGA-LUSC\n",
      "Processing data for project TCGA-LUSC\n",
      "Processing data for project TCGA-LUSC\n",
      "Processing data for project TCGA-MESO\n",
      "Processing data for project TCGA-MESO\n",
      "Processing data for project TCGA-MESO\n",
      "Processing data for project TCGA-OV\n",
      "Processing data for project TCGA-OV\n",
      "Processing data for project TCGA-OV\n",
      "Processing data for project TCGA-PAAD\n",
      "Processing data for project TCGA-PAAD\n",
      "Processing data for project TCGA-PAAD\n",
      "Processing data for project TCGA-PCPG\n",
      "Processing data for project TCGA-PCPG\n",
      "Processing data for project TCGA-PCPG\n",
      "Processing data for project TCGA-PRAD\n",
      "Processing data for project TCGA-PRAD\n",
      "Processing data for project TCGA-PRAD\n",
      "Processing data for project TCGA-READ\n",
      "Processing data for project TCGA-READ\n",
      "Processing data for project TCGA-READ\n",
      "Processing data for project TCGA-SARC\n",
      "Processing data for project TCGA-SARC\n",
      "Processing data for project TCGA-SARC\n",
      "Processing data for project TCGA-SKCM\n",
      "Processing data for project TCGA-SKCM\n",
      "Processing data for project TCGA-SKCM\n",
      "Processing data for project TCGA-STAD\n",
      "Processing data for project TCGA-STAD\n",
      "Processing data for project TCGA-STAD\n",
      "Processing data for project TCGA-TGCT\n",
      "Processing data for project TCGA-TGCT\n",
      "Processing data for project TCGA-TGCT\n",
      "Processing data for project TCGA-THCA\n",
      "Processing data for project TCGA-THCA\n",
      "Processing data for project TCGA-THCA\n",
      "Processing data for project TCGA-THYM\n",
      "Processing data for project TCGA-THYM\n",
      "Processing data for project TCGA-THYM\n",
      "Processing data for project TCGA-UCEC\n",
      "Processing data for project TCGA-UCEC\n",
      "Processing data for project TCGA-UCEC\n",
      "Processing data for project TCGA-UCS\n",
      "Processing data for project TCGA-UCS\n",
      "Processing data for project TCGA-UCS\n",
      "Processing data for project TCGA-UVM\n",
      "Processing data for project TCGA-UVM\n",
      "Processing data for project TCGA-UVM\n"
     ]
    }
   ],
   "source": [
    "for project in TCGA_PROJECTS:\n",
    "    process_project_methylation_files('DNA_Methylation_450', project)\n",
    "    process_project_methylation_files('DNA_Methylation_27', project)\n",
    "    process_project_methylation_files('DNA_Methylation', project)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Expression and Copy Number Variation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine which genes to keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gencode_df():\n",
    "    human_gencode = 'gencode.v49.annotation.gtf'\n",
    "\n",
    "    human_gencode_df = pl.read_csv(human_gencode, separator='\\t',\n",
    "                        null_values=['NA'], has_header=False, \n",
    "                        comment_prefix='#', truncate_ragged_lines=True)\n",
    "\n",
    "    human_gencode_df = human_gencode_df.rename({\n",
    "        \"column_1\": \"chromosome\",\n",
    "        \"column_2\": \"source\",\n",
    "        \"column_3\": \"feature\",\n",
    "        \"column_4\": \"start\",\n",
    "        \"column_5\": \"end\",\n",
    "        \"column_6\": \"score\",\n",
    "        \"column_7\": \"strand\",\n",
    "        \"column_8\": \"frame\",\n",
    "        \"column_9\": \"attributes\",\n",
    "    })\n",
    "\n",
    "    human_gencode_df = human_gencode_df.with_columns([\n",
    "        human_gencode_df[\"attributes\"].str.extract(r'gene_id \"([^\"]+)\"', 1).alias(\"gene_id\"),\n",
    "        human_gencode_df[\"attributes\"].str.extract(r'gene_type \"([^\"]+)\"', 1).alias(\"gene_type\"),\n",
    "    ])\n",
    "    human_gencode_df = human_gencode_df.replace_column(0,\n",
    "        pl.Series('gene_id', human_gencode_df['gene_id'].str.extract(r\"(\\w+)\"))\n",
    "    )\n",
    "\n",
    "    human_gencode_df = human_gencode_df.select([\"gene_id\", \"gene_type\"]).unique()\n",
    "    return human_gencode_df\n",
    "\n",
    "human_gencode_df = create_gencode_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_genes_to_remove_set(human_gencode_df):\n",
    "    hg38_genes_file_path = 'gencode.v36.annotation.gtf.gene.probemap' \n",
    "\n",
    "    df = pl.read_csv(hg38_genes_file_path, separator='\\t',\n",
    "                     null_values=['NA'])\n",
    "\n",
    "    selected_columns = df.select(['id', 'gene', 'chrom'])\n",
    "\n",
    "    sexual_chr_genes = set()\n",
    "    chrM_probes = set()\n",
    "    non_standard_genes = set()\n",
    "    standard_transcript_types = {\"protein_coding\", \"lncRNA\", \"miRNA\"}\n",
    "\n",
    "    gencode_map = dict(zip(human_gencode_df['gene_name'].to_list(),\n",
    "                           human_gencode_df['gene_type'].to_list()))\n",
    "\n",
    "    for row in selected_columns.iter_rows(named=True):\n",
    "        chromosome = row['chrom']  \n",
    "        gene_id = str(row['id']) \n",
    "        gene_name = str(row['gene'])\n",
    "\n",
    "        if chromosome == 'chrY' or chromosome == 'chrX':\n",
    "            sexual_chr_genes.add(gene_id)\n",
    "            continue\n",
    "\n",
    "        if chromosome=='chrM':\n",
    "            chrM_probes.add(gene_id)\n",
    "            continue\n",
    "\n",
    "        if gene_name in gencode_map.keys():\n",
    "            gene_type = gencode_map[gene_name]\n",
    "            if gene_type not in standard_transcript_types:\n",
    "                non_standard_genes.add(gene_id)\n",
    "                continue\n",
    "        \n",
    "        if '.' in gene_name:\n",
    "            non_standard_genes.add(gene_id) # Remove pseudogenes and AL/AF lncRNA\n",
    "\n",
    "    print(f\"Number of non standard genes: {len(non_standard_genes)}\")\n",
    "    print(f\"Number of sexual chromosome probes: {len(sexual_chr_genes)}\")\n",
    "    print(f\"Number of M chromosome probes: {len(chrM_probes)}\")\n",
    "\n",
    "    ensemblids_to_remove = non_standard_genes.union(sexual_chr_genes).union(chrM_probes)\n",
    "\n",
    "    return ensemblids_to_remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensemblids_to_remove = [] # create_genes_to_remove_set(human_gencode_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove unwanted genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_project_gene_level_files(project, type):\n",
    "    print(f'Processing data for project {project}')\n",
    "    if type == 'mirna':\n",
    "        project_file_path = f'UCSC_TCGA_Data/miRNA_Data/{project}'\n",
    "        output_file_path = ('UCSC_TCGA_Data_Clean/miRNA_Expression/'\n",
    "            f'{project}_filtered_mirna_expression.tsv')\n",
    "    elif type == 'gene_expression':\n",
    "        project_file_path = f'UCSC_TCGA_Data/Gene_Expression/{project}'\n",
    "        output_file_path = ('UCSC_TCGA_Data_Clean/Gene_Expression/'\n",
    "            f'{project}_filtered_gene_expression.tsv')\n",
    "    elif type == 'lncrna':\n",
    "        project_file_path = f'UCSC_TCGA_Data/Gene_Expression/{project}'\n",
    "        output_file_path = ('UCSC_TCGA_Data_Clean/lncRNA_Expression/'\n",
    "            f'{project}_filtered_lncrna_expression.tsv')\n",
    "    \n",
    "    if os.path.exists(output_file_path):\n",
    "        return\n",
    "\n",
    "    df = pl.read_csv(project_file_path, separator='\\t',\n",
    "                     null_values=['NA'])\n",
    "    print(f'Initial shape {df.shape}')  \n",
    "\n",
    "    if type == 'mirna':\n",
    "        df_filtered = df\n",
    "    elif type == 'gene_expression':\n",
    "        df = df.filter(~df['Ensembl_ID'].str.ends_with('_PAR_Y'))\n",
    "        df = df.replace_column(0,\n",
    "                pl.Series('Gene', df['Ensembl_ID'].str.extract(r\"(\\w+)\"))\n",
    "            )\n",
    "        df_filtered = df.filter(~df['Gene'].is_in(human_gencode_df['gene_id'].filter(human_gencode_df['gene_type'].is_in(['lncRNA', 'miRNA'])))\n",
    "                               )\n",
    "    elif type == 'lncrna':\n",
    "        df = df.filter(~df['Ensembl_ID'].str.ends_with('_PAR_Y'))\n",
    "        df = df.replace_column(0,\n",
    "                pl.Series('Gene', df['Ensembl_ID'].str.extract(r\"(\\w+)\"))\n",
    "            )\n",
    "        df_filtered = df.filter(df['Gene'].is_in(human_gencode_df['gene_id'].filter(human_gencode_df['gene_type'].is_in(['lncRNA']))))\n",
    "    \n",
    "    os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
    "    print(df_filtered.shape)\n",
    "    df_filtered.write_csv(output_file_path, separator='\\t')\n",
    "    print(f'Finished writing data to file {output_file_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing data for project TCGA-ACC\n",
      "Processing data for project TCGA-ACC\n",
      "Processing data for project TCGA-ACC\n",
      "Processing data for project TCGA-BLCA\n",
      "Processing data for project TCGA-BLCA\n",
      "Processing data for project TCGA-BLCA\n",
      "Processing data for project TCGA-BRCA\n",
      "Processing data for project TCGA-BRCA\n",
      "Processing data for project TCGA-BRCA\n",
      "Processing data for project TCGA-CESC\n",
      "Processing data for project TCGA-CESC\n",
      "Processing data for project TCGA-CESC\n",
      "Processing data for project TCGA-CHOL\n",
      "Processing data for project TCGA-CHOL\n",
      "Processing data for project TCGA-CHOL\n",
      "Processing data for project TCGA-COAD\n",
      "Processing data for project TCGA-COAD\n",
      "Processing data for project TCGA-COAD\n",
      "Processing data for project TCGA-DLBC\n",
      "Processing data for project TCGA-DLBC\n",
      "Processing data for project TCGA-DLBC\n",
      "Processing data for project TCGA-ESCA\n",
      "Processing data for project TCGA-ESCA\n",
      "Processing data for project TCGA-ESCA\n",
      "Processing data for project TCGA-GBM\n",
      "Processing data for project TCGA-GBM\n",
      "Processing data for project TCGA-GBM\n",
      "Processing data for project TCGA-HNSC\n",
      "Processing data for project TCGA-HNSC\n",
      "Processing data for project TCGA-HNSC\n",
      "Processing data for project TCGA-KICH\n",
      "Processing data for project TCGA-KICH\n",
      "Processing data for project TCGA-KICH\n",
      "Processing data for project TCGA-KIRC\n",
      "Processing data for project TCGA-KIRC\n",
      "Processing data for project TCGA-KIRC\n",
      "Processing data for project TCGA-KIRP\n",
      "Processing data for project TCGA-KIRP\n",
      "Processing data for project TCGA-KIRP\n",
      "Processing data for project TCGA-LAML\n",
      "Processing data for project TCGA-LAML\n",
      "Processing data for project TCGA-LAML\n",
      "Processing data for project TCGA-LGG\n",
      "Processing data for project TCGA-LGG\n",
      "Processing data for project TCGA-LGG\n",
      "Processing data for project TCGA-LIHC\n",
      "Processing data for project TCGA-LIHC\n",
      "Processing data for project TCGA-LIHC\n",
      "Processing data for project TCGA-LUAD\n",
      "Processing data for project TCGA-LUAD\n",
      "Processing data for project TCGA-LUAD\n",
      "Processing data for project TCGA-LUSC\n",
      "Processing data for project TCGA-LUSC\n",
      "Processing data for project TCGA-LUSC\n",
      "Processing data for project TCGA-MESO\n",
      "Processing data for project TCGA-MESO\n",
      "Processing data for project TCGA-MESO\n",
      "Processing data for project TCGA-OV\n",
      "Processing data for project TCGA-OV\n",
      "Processing data for project TCGA-OV\n",
      "Processing data for project TCGA-PAAD\n",
      "Processing data for project TCGA-PAAD\n",
      "Processing data for project TCGA-PAAD\n",
      "Processing data for project TCGA-PCPG\n",
      "Processing data for project TCGA-PCPG\n",
      "Processing data for project TCGA-PCPG\n",
      "Processing data for project TCGA-PRAD\n",
      "Processing data for project TCGA-PRAD\n",
      "Processing data for project TCGA-PRAD\n",
      "Processing data for project TCGA-READ\n",
      "Processing data for project TCGA-READ\n",
      "Processing data for project TCGA-READ\n",
      "Processing data for project TCGA-SARC\n",
      "Processing data for project TCGA-SARC\n",
      "Processing data for project TCGA-SARC\n",
      "Processing data for project TCGA-SKCM\n",
      "Processing data for project TCGA-SKCM\n",
      "Processing data for project TCGA-SKCM\n",
      "Processing data for project TCGA-STAD\n",
      "Processing data for project TCGA-STAD\n",
      "Processing data for project TCGA-STAD\n",
      "Processing data for project TCGA-TGCT\n",
      "Processing data for project TCGA-TGCT\n",
      "Processing data for project TCGA-TGCT\n",
      "Processing data for project TCGA-THCA\n",
      "Processing data for project TCGA-THCA\n",
      "Processing data for project TCGA-THCA\n",
      "Processing data for project TCGA-THYM\n",
      "Processing data for project TCGA-THYM\n",
      "Processing data for project TCGA-THYM\n",
      "Processing data for project TCGA-UCEC\n",
      "Processing data for project TCGA-UCEC\n",
      "Processing data for project TCGA-UCEC\n",
      "Processing data for project TCGA-UCS\n",
      "Processing data for project TCGA-UCS\n",
      "Processing data for project TCGA-UCS\n",
      "Processing data for project TCGA-UVM\n",
      "Processing data for project TCGA-UVM\n",
      "Processing data for project TCGA-UVM\n"
     ]
    }
   ],
   "source": [
    "for project in TCGA_PROJECTS:\n",
    "    process_project_gene_level_files(project, \"gene_expression\")\n",
    "    process_project_gene_level_files(project, \"lncrna\")    \n",
    "    process_project_gene_level_files(project, \"mirna\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intersect the Samples Available for each Omic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersect_project_samples(project):\n",
    "    print(f'Beginning intersection for {project}')  \n",
    "    dna_methylation_path = f'UCSC_TCGA_Data_Clean/DNA_Methylation/{project}_filtered_methylation.tsv'\n",
    "    gene_expression_path = f'UCSC_TCGA_Data_Clean/Gene_Expression/{project}_filtered_gene_expression.tsv'\n",
    "    mirna_path = f'UCSC_TCGA_Data_Clean/miRNA_Expression/{project}_filtered_mirna_expression.tsv'\n",
    "    lncrna_path = f'UCSC_TCGA_Data_Clean/lncRNA_Expression/{project}_filtered_lncrna_expression.tsv'\n",
    "\n",
    "    dna_df = pl.read_csv(dna_methylation_path, separator='\\t')\n",
    "    ge_df = pl.read_csv(gene_expression_path, separator='\\t')\n",
    "    mirna_df = pl.read_csv(mirna_path, separator='\\t')\n",
    "    lncrna_df = pl.read_csv(lncrna_path, separator='\\t')\n",
    "\n",
    "    dna_sample_cols = dna_df.select(pl.col(pl.Float64)).columns \n",
    "    ge_sample_cols = ge_df.select(pl.col(pl.Float64)).columns  \n",
    "    mirna_sample_cols = mirna_df.select(pl.col(pl.Float64)).columns\n",
    "    lncrna_sample_cols = lncrna_df.select(pl.col(pl.Float64)).columns\n",
    "    \n",
    "    first_col_dna = dna_df.columns[0]\n",
    "    first_col_ge = ge_df.columns[0]\n",
    "    first_col_mirna = mirna_df.columns[0]\n",
    "    first_col_lncrna = lncrna_df.columns[0]\n",
    "    \n",
    "    common_samples = list(set(dna_sample_cols) & set(ge_sample_cols) & set(mirna_sample_cols) & set(lncrna_sample_cols))\n",
    "    \n",
    "    dna_cols_to_keep = [first_col_dna] + common_samples\n",
    "    ge_cols_to_keep = [first_col_ge] + common_samples\n",
    "    mirna_cols_to_keep = [first_col_mirna] + common_samples\n",
    "    lncrna_cols_to_keep = [first_col_lncrna] + common_samples\n",
    "\n",
    "    dna_df_filtered = dna_df.select(dna_cols_to_keep)\n",
    "    ge_df_filtered = ge_df.select(ge_cols_to_keep)\n",
    "    mirna_df_filtered = mirna_df.select(mirna_cols_to_keep)\n",
    "    lncrna_df_filtered = lncrna_df.select(lncrna_cols_to_keep)\n",
    "    print(f'dna methylation shape: {dna_df_filtered.shape}')\n",
    "    print(f'gene expression shape: {ge_df_filtered.shape}')\n",
    "    print(f'mirna expression shape: {mirna_df_filtered.shape}')\n",
    "    print(f'lncrna expression shape: {lncrna_df_filtered.shape}')\n",
    "\n",
    "    dna_df_filtered.write_csv(dna_methylation_path, separator='\\t')\n",
    "    ge_df_filtered.write_csv(gene_expression_path, separator='\\t')\n",
    "    mirna_df_filtered.write_csv(mirna_path, separator='\\t')\n",
    "    lncrna_df_filtered.write_csv(lncrna_path, separator='\\t')\n",
    "    print(f'Finished intersecting matrices for project {project}')\n",
    "    return dna_df_filtered.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beginning intersection for TCGA-ACC\n",
      "dna methylation shape: (365468, 80)\n",
      "gene expression shape: (42964, 80)\n",
      "mirna expression shape: (1881, 80)\n",
      "lncrna expression shape: (15775, 80)\n",
      "Finished intersecting matrices for project TCGA-ACC\n",
      "Beginning intersection for TCGA-BLCA\n",
      "dna methylation shape: (365468, 424)\n",
      "gene expression shape: (42964, 424)\n",
      "mirna expression shape: (1881, 424)\n",
      "lncrna expression shape: (15775, 424)\n",
      "Finished intersecting matrices for project TCGA-BLCA\n",
      "Beginning intersection for TCGA-BRCA\n",
      "dna methylation shape: (365468, 854)\n",
      "gene expression shape: (42964, 854)\n",
      "mirna expression shape: (1881, 854)\n",
      "lncrna expression shape: (15775, 854)\n",
      "Finished intersecting matrices for project TCGA-BRCA\n",
      "Beginning intersection for TCGA-CESC\n",
      "dna methylation shape: (365468, 308)\n",
      "gene expression shape: (42964, 308)\n",
      "mirna expression shape: (1881, 308)\n",
      "lncrna expression shape: (15775, 308)\n",
      "Finished intersecting matrices for project TCGA-CESC\n",
      "Beginning intersection for TCGA-CHOL\n",
      "dna methylation shape: (365468, 45)\n",
      "gene expression shape: (42964, 45)\n",
      "mirna expression shape: (1881, 45)\n",
      "lncrna expression shape: (15775, 45)\n",
      "Finished intersecting matrices for project TCGA-CHOL\n",
      "Beginning intersection for TCGA-COAD\n",
      "dna methylation shape: (365468, 306)\n",
      "gene expression shape: (42964, 306)\n",
      "mirna expression shape: (1881, 306)\n",
      "lncrna expression shape: (15775, 306)\n",
      "Finished intersecting matrices for project TCGA-COAD\n",
      "Beginning intersection for TCGA-DLBC\n",
      "dna methylation shape: (365468, 48)\n",
      "gene expression shape: (42964, 48)\n",
      "mirna expression shape: (1881, 48)\n",
      "lncrna expression shape: (15775, 48)\n",
      "Finished intersecting matrices for project TCGA-DLBC\n",
      "Beginning intersection for TCGA-ESCA\n",
      "dna methylation shape: (365468, 195)\n",
      "gene expression shape: (42964, 195)\n",
      "mirna expression shape: (1881, 195)\n",
      "lncrna expression shape: (15775, 195)\n",
      "Finished intersecting matrices for project TCGA-ESCA\n",
      "Beginning intersection for TCGA-GBM\n",
      "dna methylation shape: (365468, 5)\n",
      "gene expression shape: (42964, 5)\n",
      "mirna expression shape: (1881, 5)\n",
      "lncrna expression shape: (15775, 5)\n",
      "Finished intersecting matrices for project TCGA-GBM\n",
      "Beginning intersection for TCGA-HNSC\n",
      "dna methylation shape: (365468, 536)\n",
      "gene expression shape: (42964, 536)\n",
      "mirna expression shape: (1881, 536)\n",
      "lncrna expression shape: (15775, 536)\n",
      "Finished intersecting matrices for project TCGA-HNSC\n",
      "Beginning intersection for TCGA-KICH\n",
      "dna methylation shape: (365468, 67)\n",
      "gene expression shape: (42964, 67)\n",
      "mirna expression shape: (1881, 67)\n",
      "lncrna expression shape: (15775, 67)\n",
      "Finished intersecting matrices for project TCGA-KICH\n",
      "Beginning intersection for TCGA-KIRC\n",
      "dna methylation shape: (365468, 346)\n",
      "gene expression shape: (42964, 346)\n",
      "mirna expression shape: (1881, 346)\n",
      "lncrna expression shape: (15775, 346)\n",
      "Finished intersecting matrices for project TCGA-KIRC\n",
      "Beginning intersection for TCGA-KIRP\n",
      "dna methylation shape: (365468, 298)\n",
      "gene expression shape: (42964, 298)\n",
      "mirna expression shape: (1881, 298)\n",
      "lncrna expression shape: (15775, 298)\n",
      "Finished intersecting matrices for project TCGA-KIRP\n",
      "Beginning intersection for TCGA-LAML\n",
      "dna methylation shape: (365468, 1)\n",
      "gene expression shape: (42964, 1)\n",
      "mirna expression shape: (1881, 1)\n",
      "lncrna expression shape: (15775, 1)\n",
      "Finished intersecting matrices for project TCGA-LAML\n",
      "Beginning intersection for TCGA-LGG\n",
      "dna methylation shape: (365468, 513)\n",
      "gene expression shape: (42964, 513)\n",
      "mirna expression shape: (1881, 513)\n",
      "lncrna expression shape: (15775, 513)\n",
      "Finished intersecting matrices for project TCGA-LGG\n",
      "Beginning intersection for TCGA-LIHC\n",
      "dna methylation shape: (365468, 409)\n",
      "gene expression shape: (42964, 409)\n",
      "mirna expression shape: (1881, 409)\n",
      "lncrna expression shape: (15775, 409)\n",
      "Finished intersecting matrices for project TCGA-LIHC\n",
      "Beginning intersection for TCGA-LUAD\n",
      "dna methylation shape: (365468, 466)\n",
      "gene expression shape: (42964, 466)\n",
      "mirna expression shape: (1881, 466)\n",
      "lncrna expression shape: (15775, 466)\n",
      "Finished intersecting matrices for project TCGA-LUAD\n",
      "Beginning intersection for TCGA-LUSC\n",
      "dna methylation shape: (365468, 369)\n",
      "gene expression shape: (42964, 369)\n",
      "mirna expression shape: (1881, 369)\n",
      "lncrna expression shape: (15775, 369)\n",
      "Finished intersecting matrices for project TCGA-LUSC\n",
      "Beginning intersection for TCGA-MESO\n",
      "dna methylation shape: (365468, 88)\n",
      "gene expression shape: (42964, 88)\n",
      "mirna expression shape: (1881, 88)\n",
      "lncrna expression shape: (15775, 88)\n",
      "Finished intersecting matrices for project TCGA-MESO\n",
      "Beginning intersection for TCGA-OV\n",
      "dna methylation shape: (365468, 10)\n",
      "gene expression shape: (42964, 10)\n",
      "mirna expression shape: (1881, 10)\n",
      "lncrna expression shape: (15775, 10)\n",
      "Finished intersecting matrices for project TCGA-OV\n",
      "Beginning intersection for TCGA-PAAD\n",
      "dna methylation shape: (365468, 183)\n",
      "gene expression shape: (42964, 183)\n",
      "mirna expression shape: (1881, 183)\n",
      "lncrna expression shape: (15775, 183)\n",
      "Finished intersecting matrices for project TCGA-PAAD\n",
      "Beginning intersection for TCGA-PCPG\n",
      "dna methylation shape: (365468, 183)\n",
      "gene expression shape: (42964, 183)\n",
      "mirna expression shape: (1881, 183)\n",
      "lncrna expression shape: (15775, 183)\n",
      "Finished intersecting matrices for project TCGA-PCPG\n",
      "Beginning intersection for TCGA-PRAD\n",
      "dna methylation shape: (365468, 533)\n",
      "gene expression shape: (42964, 533)\n",
      "mirna expression shape: (1881, 533)\n",
      "lncrna expression shape: (15775, 533)\n",
      "Finished intersecting matrices for project TCGA-PRAD\n",
      "Beginning intersection for TCGA-READ\n",
      "dna methylation shape: (365468, 99)\n",
      "gene expression shape: (42964, 99)\n",
      "mirna expression shape: (1881, 99)\n",
      "lncrna expression shape: (15775, 99)\n",
      "Finished intersecting matrices for project TCGA-READ\n",
      "Beginning intersection for TCGA-SARC\n",
      "dna methylation shape: (365468, 258)\n",
      "gene expression shape: (42964, 258)\n",
      "mirna expression shape: (1881, 258)\n",
      "lncrna expression shape: (15775, 258)\n",
      "Finished intersecting matrices for project TCGA-SARC\n",
      "Beginning intersection for TCGA-SKCM\n",
      "dna methylation shape: (365468, 99)\n",
      "gene expression shape: (42964, 99)\n",
      "mirna expression shape: (1881, 99)\n",
      "lncrna expression shape: (15775, 99)\n",
      "Finished intersecting matrices for project TCGA-SKCM\n",
      "Beginning intersection for TCGA-STAD\n",
      "dna methylation shape: (365468, 371)\n",
      "gene expression shape: (42964, 371)\n",
      "mirna expression shape: (1881, 371)\n",
      "lncrna expression shape: (15775, 371)\n",
      "Finished intersecting matrices for project TCGA-STAD\n",
      "Beginning intersection for TCGA-TGCT\n",
      "dna methylation shape: (365468, 151)\n",
      "gene expression shape: (42964, 151)\n",
      "mirna expression shape: (1881, 151)\n",
      "lncrna expression shape: (15775, 151)\n",
      "Finished intersecting matrices for project TCGA-TGCT\n",
      "Beginning intersection for TCGA-THCA\n",
      "dna methylation shape: (365468, 555)\n",
      "gene expression shape: (42964, 555)\n",
      "mirna expression shape: (1881, 555)\n",
      "lncrna expression shape: (15775, 555)\n",
      "Finished intersecting matrices for project TCGA-THCA\n",
      "Beginning intersection for TCGA-THYM\n",
      "dna methylation shape: (365468, 123)\n",
      "gene expression shape: (42964, 123)\n",
      "mirna expression shape: (1881, 123)\n",
      "lncrna expression shape: (15775, 123)\n",
      "Finished intersecting matrices for project TCGA-THYM\n",
      "Beginning intersection for TCGA-UCEC\n",
      "dna methylation shape: (365468, 465)\n",
      "gene expression shape: (42964, 465)\n",
      "mirna expression shape: (1881, 465)\n",
      "lncrna expression shape: (15775, 465)\n",
      "Finished intersecting matrices for project TCGA-UCEC\n",
      "Beginning intersection for TCGA-UCS\n",
      "dna methylation shape: (365468, 58)\n",
      "gene expression shape: (42964, 58)\n",
      "mirna expression shape: (1881, 58)\n",
      "lncrna expression shape: (15775, 58)\n",
      "Finished intersecting matrices for project TCGA-UCS\n",
      "Beginning intersection for TCGA-UVM\n",
      "dna methylation shape: (365468, 81)\n",
      "gene expression shape: (42964, 81)\n",
      "mirna expression shape: (1881, 81)\n",
      "lncrna expression shape: (15775, 81)\n",
      "Finished intersecting matrices for project TCGA-UVM\n",
      "Total samples: 8527\n"
     ]
    }
   ],
   "source": [
    "sample_number = 0\n",
    "for project in TCGA_PROJECTS:\n",
    "    count = intersect_project_samples(project)\n",
    "    sample_number += int(count)\n",
    "print(f'Total samples: {sample_number}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verify Sample Type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove samples that are not primary tumor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_non_primary_tumor_samples(project):\n",
    "    \n",
    "    dna_methylation_path = f'UCSC_TCGA_Data_Clean/DNA_Methylation/{project}_filtered_methylation.tsv'\n",
    "    gene_expression_path = f'UCSC_TCGA_Data_Clean/Gene_Expression/{project}_filtered_gene_expression.tsv'\n",
    "    mirna_expression_path = f'UCSC_TCGA_Data_Clean/miRNA_Expression/{project}_filtered_mirna_expression.tsv'\n",
    "    lncrna_expression_path = f'UCSC_TCGA_Data_Clean/lncRNA_Expression/{project}_filtered_lncrna_expression.tsv'\n",
    "\n",
    "    dna_df = pl.read_csv(dna_methylation_path, separator='\\t')\n",
    "    ge_df = pl.read_csv(gene_expression_path, separator='\\t')\n",
    "    mirna_df = pl.read_csv(mirna_expression_path, separator='\\t')\n",
    "    lncrna_df = pl.read_csv(lncrna_expression_path, separator='\\t')\n",
    "    print(f'Initial number of samples {dna_df.shape[1]}')\n",
    "\n",
    "    first_col_dna = dna_df.columns[0]\n",
    "    first_col_ge = ge_df.columns[0]\n",
    "    first_col_mirna = mirna_df.columns[0]\n",
    "    first_col_lncrna = lncrna_df.columns[0]\n",
    "\n",
    "    project_sample_types = {}\n",
    "\n",
    "    columns_to_keep = []  \n",
    "    dna_sample_cols = dna_df.select(pl.col(pl.Float64)).columns \n",
    "    for column in dna_sample_cols:\n",
    "        sample_type = column.split('-')[3][:2] \n",
    "        if sample_type not in project_sample_types:\n",
    "            project_sample_types[sample_type] = 1\n",
    "        else:\n",
    "            project_sample_types[sample_type] += 1\n",
    "        if sample_type == '01' or sample_type == '11':   \n",
    "            columns_to_keep.append(column)\n",
    "\n",
    "    print(project_sample_types)\n",
    "\n",
    "    dna_cols_to_keep = [first_col_dna] + columns_to_keep\n",
    "    ge_cols_to_keep = [first_col_ge] + columns_to_keep\n",
    "    mirna_cols_to_keep = [first_col_mirna] + columns_to_keep\n",
    "    lncrna_cols_to_keep = [first_col_lncrna] + columns_to_keep\n",
    "\n",
    "    dna_df_filtered = dna_df.select(dna_cols_to_keep)\n",
    "    ge_df_filtered = ge_df.select(ge_cols_to_keep)\n",
    "    mirna_df_filtered = mirna_df.select(mirna_cols_to_keep)\n",
    "    lncrna_df_filtered = lncrna_df.select(lncrna_cols_to_keep)\n",
    "\n",
    "    dna_df_filtered.write_csv(dna_methylation_path, separator='\\t')\n",
    "    ge_df_filtered.write_csv(gene_expression_path, separator='\\t')\n",
    "    mirna_df_filtered.write_csv(mirna_expression_path, separator='\\t')\n",
    "    lncrna_df_filtered.write_csv(lncrna_expression_path, separator='\\t')\n",
    "    print(f'Final number of samples {dna_df_filtered.shape[1]}')\n",
    "    print(f'Finished verifying samples type for project {project}')\n",
    "    return dna_df_filtered.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial number of samples 80\n",
      "{'01': 79}\n",
      "Final number of samples 80\n",
      "Finished verifying samples type for project TCGA-ACC\n",
      "Initial number of samples 424\n",
      "{'01': 406, '11': 17}\n",
      "Final number of samples 424\n",
      "Finished verifying samples type for project TCGA-BLCA\n",
      "Initial number of samples 854\n",
      "{'01': 777, '11': 76}\n",
      "Final number of samples 854\n",
      "Finished verifying samples type for project TCGA-BRCA\n",
      "Initial number of samples 308\n",
      "{'01': 304, '11': 3}\n",
      "Final number of samples 308\n",
      "Finished verifying samples type for project TCGA-CESC\n",
      "Initial number of samples 45\n",
      "{'11': 9, '01': 35}\n",
      "Final number of samples 45\n",
      "Finished verifying samples type for project TCGA-CHOL\n",
      "Initial number of samples 306\n",
      "{'01': 301, '11': 4}\n",
      "Final number of samples 306\n",
      "Finished verifying samples type for project TCGA-COAD\n",
      "Initial number of samples 48\n",
      "{'01': 47}\n",
      "Final number of samples 48\n",
      "Finished verifying samples type for project TCGA-DLBC\n",
      "Initial number of samples 195\n",
      "{'01': 183, '11': 11}\n",
      "Final number of samples 195\n",
      "Finished verifying samples type for project TCGA-ESCA\n",
      "Initial number of samples 5\n",
      "{'01': 3, '11': 1}\n",
      "Final number of samples 5\n",
      "Finished verifying samples type for project TCGA-GBM\n",
      "Initial number of samples 536\n",
      "{'01': 515, '11': 20}\n",
      "Final number of samples 536\n",
      "Finished verifying samples type for project TCGA-HNSC\n",
      "Initial number of samples 67\n",
      "{'01': 66}\n",
      "Final number of samples 67\n",
      "Finished verifying samples type for project TCGA-KICH\n",
      "Initial number of samples 346\n",
      "{'01': 321, '11': 24}\n",
      "Final number of samples 346\n",
      "Finished verifying samples type for project TCGA-KIRC\n",
      "Initial number of samples 298\n",
      "{'01': 274, '11': 23}\n",
      "Final number of samples 298\n",
      "Finished verifying samples type for project TCGA-KIRP\n",
      "Initial number of samples 1\n",
      "{}\n",
      "Final number of samples 1\n",
      "Finished verifying samples type for project TCGA-LAML\n",
      "Initial number of samples 513\n",
      "{'01': 512}\n",
      "Final number of samples 513\n",
      "Finished verifying samples type for project TCGA-LGG\n",
      "Initial number of samples 409\n",
      "{'01': 367, '11': 41}\n",
      "Final number of samples 409\n",
      "Finished verifying samples type for project TCGA-LIHC\n",
      "Initial number of samples 466\n",
      "{'01': 456, '11': 9}\n",
      "Final number of samples 466\n",
      "Finished verifying samples type for project TCGA-LUAD\n",
      "Initial number of samples 369\n",
      "{'01': 365, '11': 3}\n",
      "Final number of samples 369\n",
      "Finished verifying samples type for project TCGA-LUSC\n",
      "Initial number of samples 88\n",
      "{'01': 87}\n",
      "Final number of samples 88\n",
      "Finished verifying samples type for project TCGA-MESO\n",
      "Initial number of samples 10\n",
      "{'01': 9}\n",
      "Final number of samples 10\n",
      "Finished verifying samples type for project TCGA-OV\n",
      "Initial number of samples 183\n",
      "{'01': 178, '11': 4}\n",
      "Final number of samples 183\n",
      "Finished verifying samples type for project TCGA-PAAD\n",
      "Initial number of samples 183\n",
      "{'01': 179, '11': 3}\n",
      "Final number of samples 183\n",
      "Finished verifying samples type for project TCGA-PCPG\n",
      "Initial number of samples 533\n",
      "{'11': 35, '01': 497}\n",
      "Final number of samples 533\n",
      "Finished verifying samples type for project TCGA-PRAD\n",
      "Initial number of samples 99\n",
      "{'01': 98}\n",
      "Final number of samples 99\n",
      "Finished verifying samples type for project TCGA-READ\n",
      "Initial number of samples 258\n",
      "{'01': 257}\n",
      "Final number of samples 258\n",
      "Finished verifying samples type for project TCGA-SARC\n",
      "Initial number of samples 99\n",
      "{'01': 97, '11': 1}\n",
      "Final number of samples 99\n",
      "Finished verifying samples type for project TCGA-SKCM\n",
      "Initial number of samples 371\n",
      "{'01': 370}\n",
      "Final number of samples 371\n",
      "Finished verifying samples type for project TCGA-STAD\n",
      "Initial number of samples 151\n",
      "{'01': 150}\n",
      "Final number of samples 151\n",
      "Finished verifying samples type for project TCGA-TGCT\n",
      "Initial number of samples 555\n",
      "{'01': 504, '11': 50}\n",
      "Final number of samples 555\n",
      "Finished verifying samples type for project TCGA-THCA\n",
      "Initial number of samples 123\n",
      "{'01': 120, '11': 2}\n",
      "Final number of samples 123\n",
      "Finished verifying samples type for project TCGA-THYM\n",
      "Initial number of samples 465\n",
      "{'01': 432, '11': 32}\n",
      "Final number of samples 465\n",
      "Finished verifying samples type for project TCGA-UCEC\n",
      "Initial number of samples 58\n",
      "{'01': 57}\n",
      "Final number of samples 58\n",
      "Finished verifying samples type for project TCGA-UCS\n",
      "Initial number of samples 81\n",
      "{'01': 80}\n",
      "Final number of samples 81\n",
      "Finished verifying samples type for project TCGA-UVM\n",
      "Total samples: 8527\n"
     ]
    }
   ],
   "source": [
    "sample_number = 0\n",
    "for project in TCGA_PROJECTS:\n",
    "    count = remove_non_primary_tumor_samples(project)\n",
    "    sample_number += count\n",
    "print(f'Total samples: {sample_number}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Replace NaN for null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_all_nan():\n",
    "    directories = ['DNA_Methylation', 'Gene_Expression', 'miRNA_Expression', 'lncRNA_Expression']\n",
    "    for directory in directories:\n",
    "        if not Path('UCSC_TCGA_Data_Clean').joinpath(directory).exists():\n",
    "            continue\n",
    "        for file_path in Path('UCSC_TCGA_Data_Clean').joinpath(directory).glob('**/*.tsv'):\n",
    "            df = pl.read_csv(file_path, separator='\\t')\n",
    "            print((df.null_count().sum_horizontal() > 0).any())\n",
    "            replaced_nan_df = df.fill_nan(None)\n",
    "            print((replaced_nan_df.null_count().sum_horizontal() > 0).any())\n",
    "            replaced_nan_df.write_csv(file_path, separator='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "False\n",
      "False\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "replace_all_nan()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aggregate DataFrames by omics and remove Missing Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign the maximum number of features after the previously executed\n",
    "# it is hard coded but can also be simply augmented dynamically by reading the file of the first project\n",
    "MAX_FEATURE_NUMBER = {\n",
    "    'miRNA_Expression': 1881,\n",
    "    'lncRNA_Expression': 15786,\n",
    "    'Gene_Expression': 42995\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_size(obj, seen=None):\n",
    "    \"\"\"Recursively finds size of objects\"\"\"\n",
    "    size = sys.getsizeof(obj)\n",
    "    if seen is None:\n",
    "        seen = set()\n",
    "    obj_id = id(obj)\n",
    "    if obj_id in seen:\n",
    "        return 0\n",
    "    # Important mark as seen *before* entering recursion to gracefully handle\n",
    "    # self-referential objects\n",
    "    seen.add(obj_id)\n",
    "    if isinstance(obj, dict):\n",
    "        size += sum([get_size(v, seen) for v in obj.values()])\n",
    "        size += sum([get_size(k, seen) for k in obj.keys()])\n",
    "    elif hasattr(obj, '__dict__'):\n",
    "        size += get_size(obj.__dict__, seen)\n",
    "    elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes, bytearray)):\n",
    "        size += sum([get_size(i, seen) for i in obj])\n",
    "    return size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-01-15 12:28:26.214965: Reading all patient ids\n",
      "2026-01-15 12:28:52.009226: All patient ids read, in total 8494 patients\n"
     ]
    }
   ],
   "source": [
    "def get_all_patients(omics_type):\n",
    "    print(f'{dt.datetime.now()}: Reading all patient ids')\n",
    "    directory = f'UCSC_TCGA_Data_Clean/{omics_type}'\n",
    "    columns = []\n",
    "    for file in os.listdir(directory):\n",
    "        file_path = os.path.join(directory, file)\n",
    "        \n",
    "        if not os.path.isdir(file_path):\n",
    "            df = pl.read_csv(file_path, separator='\\t', null_values=['NA'])\n",
    "            columns.extend([*df.columns[1:]])\n",
    "    print(f'{dt.datetime.now()}: All patient ids read, in total {len(columns)} patients')\n",
    "    return columns\n",
    "\n",
    "# drop the patients from the illumina 27 list which were filtered by the intersection method above\n",
    "all_patients = get_all_patients('DNA_Methylation')\n",
    "illumina_27_patients = [patient for patient in illumina_27_patients if patient in all_patients]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_missing_values_threshold(df, var): \n",
    "    numeric_cols = df.columns[1:] \n",
    "    num_columns = len(numeric_cols)\n",
    "    na_threshold = num_columns * 0.05\n",
    "    variance_threshold = var\n",
    "\n",
    "    filtered_df = df.with_columns(\n",
    "            pl.sum_horizontal([pl.col(c).is_null() for c in numeric_cols])\n",
    "                .alias(\"na_count\"),\n",
    "            pl.concat_list([pl.col(c).cast(pl.Float64) for c in numeric_cols])\n",
    "            .list.var()\n",
    "            .alias(\"row_var\")\n",
    "        )\n",
    "    if var != 0:\n",
    "            filtered_df = filtered_df.filter(\n",
    "                (pl.col(\"na_count\") <= na_threshold) &\n",
    "                (pl.col(\"row_var\") >= variance_threshold) \n",
    "            ).drop([\"na_count\", \"row_var\"]) \n",
    "    else: \n",
    "        filtered_df = filtered_df.filter(\n",
    "                (pl.col(\"na_count\") <= na_threshold) &\n",
    "                (pl.col(\"row_var\") > variance_threshold)\n",
    "            ).drop([\"na_count\", 'row_var'])\n",
    "\n",
    "    print(filtered_df.shape)\n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_missing_and_zero_expression(type, id_col, chunks):\n",
    "    directory = f'UCSC_TCGA_Data_Clean/{type}'\n",
    "    output_directory = f'UCSC_TCGA_Data_Clean/Aggregated/'\n",
    "    output_file_path = os.path.join(output_directory, f'{type}.tsv')\n",
    "    patients = [id_col]\n",
    "    str_cols = {id_col}\n",
    "    patients.extend(get_all_patients(type))\n",
    "    out_df_cols = {col: [0.0] for col in patients if col not in str_cols}\n",
    "    out_df_cols[id_col] = ['']\n",
    "    out_df = pl.DataFrame(out_df_cols)\n",
    "    out_df = out_df.select(patients)\n",
    "    chunk_size = int(MAX_FEATURE_NUMBER[type] / chunks + 1)\n",
    "    for i in range(chunks):\n",
    "        print(f'{dt.datetime.now()}: Start run {i} to filter the {type}')\n",
    "        # load the data chunk by chunk as pandas dataframe\n",
    "        agg_df = pd.DataFrame()\n",
    "        for file in os.listdir(directory):\n",
    "            file_path = os.path.join(directory, file)\n",
    "        \n",
    "            if not os.path.isdir(file_path):\n",
    "                df = pd.read_csv(file_path, sep='\\t', index_col=0)\n",
    "                df = df.iloc[i * chunk_size: (i + 1) * chunk_size, :]\n",
    "                agg_df = df.join(agg_df, on=id_col, how='left')\n",
    "\n",
    "        # perform a numpy calculation to convert the log odds data back to normalized counts\n",
    "        # there does not exist any numpy interface with polars, such that a pandas dataframe is required\n",
    "        agg_df = np.exp2(agg_df) - 1\n",
    "        # convert the dataframe to a polars dataframe and proceed\n",
    "        agg_index = pl.from_pandas(agg_df.index)\n",
    "        agg_df = pl.from_pandas(agg_df)\n",
    "        agg_df = agg_df.insert_column(0, agg_index)\n",
    "\n",
    "        print(f'{dt.datetime.now()}: The aggregated DataFrame has shape {agg_df.shape}')\n",
    "        filtered_df = remove_missing_values_threshold(agg_df, 0.05)\n",
    "        out_df = out_df.extend(filtered_df.select(patients))\n",
    "\n",
    "    print(f'{dt.datetime.now()}: Writing the filtered and aggregated file with dimension {out_df.shape}')\n",
    "    os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
    "    print(filtered_df.shape)\n",
    "    filtered_df.write_csv(output_file_path, separator='\\t')\n",
    "    print(f'Finished writing data to file {output_file_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_missing_values_methylation():\n",
    "    directory = f'UCSC_TCGA_Data_Clean/DNA_Methylation'\n",
    "    output_directory = f'UCSC_TCGA_Data_Clean/Aggregated'\n",
    "    output_file_path = os.path.join(output_directory, 'DNA_Methylation.tsv')\n",
    "    patients = ['Probe']\n",
    "    str_cols = {'Probe'}\n",
    "    patients.extend(get_all_patients('DNA_Methylation'))\n",
    "    out_df_cols = {col: [0.0] for col in patients if col not in str_cols}\n",
    "    out_df_cols['Probe'] = ['']\n",
    "    out_df = pl.DataFrame(out_df_cols)\n",
    "    out_df = out_df.select(patients)\n",
    "    # write the header to the output file\n",
    "    os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
    "    out_df.write_csv(output_file_path, separator='\\t')\n",
    "    # define the number of chunks to reduce the amount of used RAM\n",
    "    chunks = 32\n",
    "    chunk_size = int(len(probes_to_keep) / chunks + 1)\n",
    "    for i in range(chunks):\n",
    "        print(f'{dt.datetime.now()}: Start run {i} to filter the DNA methylation sides')\n",
    "        agg_df = pl.DataFrame({'Probe': ['']})\n",
    "        for file in os.listdir(directory):\n",
    "            # output_file_path = os.path.join(output_directory, file.split('.')[0] + '_clean.tsv')\n",
    "            file_path = os.path.join(directory, file)\n",
    "        \n",
    "            if not os.path.isdir(file_path):\n",
    "                df = pl.read_csv(file_path, separator='\\t', null_values=['NA']).slice(i * chunk_size, chunk_size)\n",
    "                agg_df = df.join(agg_df, on='Probe', how='left')\n",
    "\n",
    "        \n",
    "        print(f'{dt.datetime.now()}: The aggregated DataFrame has shape {agg_df.shape}')\n",
    "        # filtered_450_df = remove_missing_values_threshold(agg_df.drop(illumina_27_patients), 0.05)\n",
    "        # filtered_27_df = remove_missing_values_threshold(agg_df.select(['Probe'] + illumina_27_patients), 0.05)\n",
    "        # filtered_df = filtered_27_df.join(filtered_450_df, on='Probe', how='full', coalesce=True)\n",
    "        filtered_df = remove_missing_values_threshold(agg_df.drop(illumina_27_patients), 0.05)\n",
    "        out_df = filtered_df.select(patients)\n",
    "        with open(output_file_path, mode='a') as f:\n",
    "            print(f'{dt.datetime.now()}: Writing the filtered data of chunk {i} with dimension {out_df.shape}')\n",
    "            out_df.write_csv(f, separator='\\t', include_header=False)\n",
    "            print('Data has been appended')\n",
    "\n",
    "    print(f'Finished writing data to file {output_file_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-01-15 12:28:52.102654: Reading all patient ids\n",
      "2026-01-15 12:28:52.977058: All patient ids read, in total 8494 patients\n",
      "2026-01-15 12:28:53.056492: Start run 0 to filter the miRNA_Expression\n",
      "2026-01-15 12:28:56.466318: The aggregated DataFrame has shape (1881, 8495)\n",
      "(932, 8495)\n",
      "2026-01-15 12:28:58.560536: Writing the filtered and aggregated file with dimension (933, 8495)\n",
      "(932, 8495)\n",
      "Finished writing data to file UCSC_TCGA_Data_Clean/Aggregated/miRNA_Expression.tsv\n",
      "2026-01-15 12:28:59.203344: Reading all patient ids\n",
      "2026-01-15 12:29:00.711052: All patient ids read, in total 8494 patients\n",
      "2026-01-15 12:29:00.780316: Start run 0 to filter the lncRNA_Expression\n",
      "2026-01-15 12:29:25.086301: The aggregated DataFrame has shape (15775, 8495)\n",
      "(7450, 8495)\n",
      "2026-01-15 12:29:41.127061: Writing the filtered and aggregated file with dimension (7451, 8495)\n",
      "(7450, 8495)\n",
      "Finished writing data to file UCSC_TCGA_Data_Clean/Aggregated/lncRNA_Expression.tsv\n",
      "2026-01-15 12:29:43.655315: Reading all patient ids\n",
      "2026-01-15 12:29:46.286502: All patient ids read, in total 8494 patients\n",
      "2026-01-15 12:29:46.340997: Start run 0 to filter the Gene_Expression\n",
      "2026-01-15 12:31:49.893124: The aggregated DataFrame has shape (42964, 8495)\n",
      "(25836, 8495)\n",
      "2026-01-15 12:32:34.722994: Writing the filtered and aggregated file with dimension (25837, 8495)\n",
      "(25836, 8495)\n",
      "Finished writing data to file UCSC_TCGA_Data_Clean/Aggregated/Gene_Expression.tsv\n"
     ]
    }
   ],
   "source": [
    "remove_missing_and_zero_expression('miRNA_Expression', 'miRNA_ID', 1)\n",
    "remove_missing_and_zero_expression('lncRNA_Expression', 'Gene', 1)\n",
    "remove_missing_and_zero_expression('Gene_Expression', 'Gene', 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def join_separate_methylation_protocols():\n",
    "    # the protocols are processed separately to strictly filter the data\n",
    "    # As a lot of patient data processed with Illumina 27k protocol contains 450.000 zeros, the variance is more likely to have a higher variance\n",
    "    # To avoid filtering just because the protocol it is filtered by two different protocols, the filtering procedure is done seperately\n",
    "    print(f'Reading and joining both DNA Methylation files for project {project}')\n",
    "    df_27 = pl.read_csv(file_path_27, separator='\\t')\n",
    "    df_450 = pl.read_csv(file_path_450, separator='\\t')\n",
    "    \n",
    "    df = df_27.join(df_450, on='Composite Element REF', how='full', coalesce=True)\n",
    "    print(f'Files joined from shapes {df_27.shape} and {df_450.shape} to shape {df.shape}')\n",
    "        \n",
    "    df.write_csv(output_file, separator='\\t')\n",
    "    print(f'Methylation files joined into {output_file}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-01-15 12:32:41.459765: Reading all patient ids\n",
      "2026-01-15 12:33:02.058517: All patient ids read, in total 8494 patients\n",
      "2026-01-15 12:33:02.152039: Start run 0 to filter the DNA methylation sides\n",
      "2026-01-15 12:33:31.601682: The aggregated DataFrame has shape (11395, 8495)\n",
      "(752, 8495)\n",
      "2026-01-15 12:33:45.916020: Writing the filtered data of chunk 0 with dimension (752, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:33:46.415193: Start run 1 to filter the DNA methylation sides\n",
      "2026-01-15 12:34:13.591039: The aggregated DataFrame has shape (11395, 8495)\n",
      "(748, 8495)\n",
      "2026-01-15 12:34:28.028348: Writing the filtered data of chunk 1 with dimension (748, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:34:28.470694: Start run 2 to filter the DNA methylation sides\n",
      "2026-01-15 12:34:50.992502: The aggregated DataFrame has shape (11395, 8495)\n",
      "(771, 8495)\n",
      "2026-01-15 12:35:05.831221: Writing the filtered data of chunk 2 with dimension (771, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:35:06.222218: Start run 3 to filter the DNA methylation sides\n",
      "2026-01-15 12:35:33.434472: The aggregated DataFrame has shape (11395, 8495)\n",
      "(795, 8495)\n",
      "2026-01-15 12:35:47.956122: Writing the filtered data of chunk 3 with dimension (795, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:35:48.451142: Start run 4 to filter the DNA methylation sides\n",
      "2026-01-15 12:36:10.564184: The aggregated DataFrame has shape (11395, 8495)\n",
      "(786, 8495)\n",
      "2026-01-15 12:36:25.413874: Writing the filtered data of chunk 4 with dimension (786, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:36:25.851097: Start run 5 to filter the DNA methylation sides\n",
      "2026-01-15 12:36:47.491958: The aggregated DataFrame has shape (11395, 8495)\n",
      "(767, 8495)\n",
      "2026-01-15 12:37:02.349591: Writing the filtered data of chunk 5 with dimension (767, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:37:02.855755: Start run 6 to filter the DNA methylation sides\n",
      "2026-01-15 12:37:26.075840: The aggregated DataFrame has shape (11395, 8495)\n",
      "(761, 8495)\n",
      "2026-01-15 12:37:40.818269: Writing the filtered data of chunk 6 with dimension (761, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:37:41.288893: Start run 7 to filter the DNA methylation sides\n",
      "2026-01-15 12:38:05.576395: The aggregated DataFrame has shape (11395, 8495)\n",
      "(804, 8495)\n",
      "2026-01-15 12:38:20.187394: Writing the filtered data of chunk 7 with dimension (804, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:38:20.684240: Start run 8 to filter the DNA methylation sides\n",
      "2026-01-15 12:38:43.225558: The aggregated DataFrame has shape (11395, 8495)\n",
      "(796, 8495)\n",
      "2026-01-15 12:38:58.071416: Writing the filtered data of chunk 8 with dimension (796, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:38:58.555185: Start run 9 to filter the DNA methylation sides\n",
      "2026-01-15 12:39:26.919535: The aggregated DataFrame has shape (11395, 8495)\n",
      "(733, 8495)\n",
      "2026-01-15 12:39:41.871736: Writing the filtered data of chunk 9 with dimension (733, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:39:42.388289: Start run 10 to filter the DNA methylation sides\n",
      "2026-01-15 12:40:07.226036: The aggregated DataFrame has shape (11395, 8495)\n",
      "(864, 8495)\n",
      "2026-01-15 12:40:21.949621: Writing the filtered data of chunk 10 with dimension (864, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:40:22.461957: Start run 11 to filter the DNA methylation sides\n",
      "2026-01-15 12:40:45.524952: The aggregated DataFrame has shape (11395, 8495)\n",
      "(829, 8495)\n",
      "2026-01-15 12:41:00.553086: Writing the filtered data of chunk 11 with dimension (829, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:41:01.016812: Start run 12 to filter the DNA methylation sides\n",
      "2026-01-15 12:41:24.944089: The aggregated DataFrame has shape (11395, 8495)\n",
      "(800, 8495)\n",
      "2026-01-15 12:41:41.808191: Writing the filtered data of chunk 12 with dimension (800, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:41:42.193909: Start run 13 to filter the DNA methylation sides\n",
      "2026-01-15 12:42:05.916205: The aggregated DataFrame has shape (11395, 8495)\n",
      "(787, 8495)\n",
      "2026-01-15 12:42:21.266410: Writing the filtered data of chunk 13 with dimension (787, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:42:21.630865: Start run 14 to filter the DNA methylation sides\n",
      "2026-01-15 12:42:45.881036: The aggregated DataFrame has shape (11395, 8495)\n",
      "(783, 8495)\n",
      "2026-01-15 12:43:01.425823: Writing the filtered data of chunk 14 with dimension (783, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:43:01.863288: Start run 15 to filter the DNA methylation sides\n",
      "2026-01-15 12:43:30.869206: The aggregated DataFrame has shape (11395, 8495)\n",
      "(843, 8495)\n",
      "2026-01-15 12:43:46.475022: Writing the filtered data of chunk 15 with dimension (843, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:43:46.843463: Start run 16 to filter the DNA methylation sides\n",
      "2026-01-15 12:44:12.236843: The aggregated DataFrame has shape (11395, 8495)\n",
      "(781, 8495)\n",
      "2026-01-15 12:44:27.251281: Writing the filtered data of chunk 16 with dimension (781, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:44:27.801716: Start run 17 to filter the DNA methylation sides\n",
      "2026-01-15 12:44:57.497647: The aggregated DataFrame has shape (11395, 8495)\n",
      "(759, 8495)\n",
      "2026-01-15 12:45:12.226312: Writing the filtered data of chunk 17 with dimension (759, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:45:12.624391: Start run 18 to filter the DNA methylation sides\n",
      "2026-01-15 12:45:36.267243: The aggregated DataFrame has shape (11395, 8495)\n",
      "(800, 8495)\n",
      "2026-01-15 12:45:50.760211: Writing the filtered data of chunk 18 with dimension (800, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:45:51.122940: Start run 19 to filter the DNA methylation sides\n",
      "2026-01-15 12:46:15.724663: The aggregated DataFrame has shape (11395, 8495)\n",
      "(792, 8495)\n",
      "2026-01-15 12:46:31.695060: Writing the filtered data of chunk 19 with dimension (792, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:46:32.143317: Start run 20 to filter the DNA methylation sides\n",
      "2026-01-15 12:46:55.604013: The aggregated DataFrame has shape (11395, 8495)\n",
      "(783, 8495)\n",
      "2026-01-15 12:47:10.828565: Writing the filtered data of chunk 20 with dimension (783, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:47:11.259442: Start run 21 to filter the DNA methylation sides\n",
      "2026-01-15 12:47:34.651106: The aggregated DataFrame has shape (11395, 8495)\n",
      "(844, 8495)\n",
      "2026-01-15 12:47:48.886532: Writing the filtered data of chunk 21 with dimension (844, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:47:49.337664: Start run 22 to filter the DNA methylation sides\n",
      "2026-01-15 12:48:13.391614: The aggregated DataFrame has shape (11395, 8495)\n",
      "(772, 8495)\n",
      "2026-01-15 12:48:28.083437: Writing the filtered data of chunk 22 with dimension (772, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:48:28.435375: Start run 23 to filter the DNA methylation sides\n",
      "2026-01-15 12:48:52.856141: The aggregated DataFrame has shape (11395, 8495)\n",
      "(812, 8495)\n",
      "2026-01-15 12:49:08.893897: Writing the filtered data of chunk 23 with dimension (812, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:49:09.390743: Start run 24 to filter the DNA methylation sides\n",
      "2026-01-15 12:49:33.179334: The aggregated DataFrame has shape (11395, 8495)\n",
      "(764, 8495)\n",
      "2026-01-15 12:49:50.345760: Writing the filtered data of chunk 24 with dimension (764, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:49:50.796007: Start run 25 to filter the DNA methylation sides\n",
      "2026-01-15 12:50:13.255104: The aggregated DataFrame has shape (11395, 8495)\n",
      "(798, 8495)\n",
      "2026-01-15 12:50:28.354128: Writing the filtered data of chunk 25 with dimension (798, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:50:28.873038: Start run 26 to filter the DNA methylation sides\n",
      "2026-01-15 12:50:54.308744: The aggregated DataFrame has shape (11395, 8495)\n",
      "(796, 8495)\n",
      "2026-01-15 12:51:09.967386: Writing the filtered data of chunk 26 with dimension (796, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:51:10.439711: Start run 27 to filter the DNA methylation sides\n",
      "2026-01-15 12:51:38.022445: The aggregated DataFrame has shape (11395, 8495)\n",
      "(838, 8495)\n",
      "2026-01-15 12:51:52.878224: Writing the filtered data of chunk 27 with dimension (838, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:51:53.283178: Start run 28 to filter the DNA methylation sides\n",
      "2026-01-15 12:52:18.319051: The aggregated DataFrame has shape (11395, 8495)\n",
      "(831, 8495)\n",
      "2026-01-15 12:52:33.637688: Writing the filtered data of chunk 28 with dimension (831, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:52:34.099504: Start run 29 to filter the DNA methylation sides\n",
      "2026-01-15 12:52:58.495920: The aggregated DataFrame has shape (11395, 8495)\n",
      "(810, 8495)\n",
      "2026-01-15 12:53:12.664910: Writing the filtered data of chunk 29 with dimension (810, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:53:13.050409: Start run 30 to filter the DNA methylation sides\n",
      "2026-01-15 12:53:37.552568: The aggregated DataFrame has shape (11395, 8495)\n",
      "(856, 8495)\n",
      "2026-01-15 12:53:51.922106: Writing the filtered data of chunk 30 with dimension (856, 8495)\n",
      "Data has been appended\n",
      "2026-01-15 12:53:52.365592: Start run 31 to filter the DNA methylation sides\n",
      "2026-01-15 12:54:17.271290: The aggregated DataFrame has shape (11395, 8495)\n",
      "(717, 8495)\n",
      "2026-01-15 12:54:31.350871: Writing the filtered data of chunk 31 with dimension (717, 8495)\n",
      "Data has been appended\n",
      "Finished writing data to file UCSC_TCGA_Data_Clean/Aggregated/DNA_Methylation.tsv\n"
     ]
    }
   ],
   "source": [
    "remove_missing_values_methylation()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove Unneeded Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_files_in_directory(dir):\n",
    "    for file in os.listdir(dir):\n",
    "        file_path = os.path.join(dir, file)\n",
    "        os.remove(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Free up device space\n",
    "directories = [\n",
    "    'UCSC_TCGA_Data/Gene_Expression',\n",
    "    'UCSC_TCGA_Data/DNA_Methylation',\n",
    "    # 'UCSC_TCGA_Data/DNA_Methylation_27',\n",
    "    'UCSC_TCGA_Data/DNA_Methylation_450',\n",
    "    'UCSC_TCGA_Data/miRNA_Data',\n",
    "]\n",
    "\n",
    "for dir in directories:\n",
    "    remove_files_in_directory(dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Clinical Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for project in TCGA_PROJECTS:\n",
    "    download_tcga_data(project, \"clinical\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_clinical_data_df():\n",
    "    directory = 'UCSC_TCGA_Data/Clinical_Data/'\n",
    "    output_file_path = 'UCSC_TCGA_Data_Clean/clinical.tsv'\n",
    "    barcode_project_list = []\n",
    "    survival_time_column = ['days_to_last_follow_up.diagnoses', 'days_to_death.demographic']\n",
    "    if len(os.listdir('UCSC_TCGA_Data_Clean/')) > 0:\n",
    "        tcga_barcodes = set(pl.read_csv('UCSC_TCGA_Data_Clean/Aggregated/miRNA_Expression.tsv', separator='\\t').columns[1:])\n",
    "    \n",
    "    for file in os.listdir(directory):\n",
    "        file_path = os.path.join(directory, file)\n",
    "        df = pl.read_csv(file_path, separator='\\t', schema_overrides={'submitter_id.annotations': pl.Utf8()},ignore_errors=True)\n",
    "\n",
    "        selected_columns = df.select(['sample', 'project_id.project', 'tissue_type.samples', 'vital_status.demographic',\n",
    "                                      'days_to_death.demographic', 'days_to_last_follow_up.diagnoses'])\n",
    "\n",
    "        for row in selected_columns.iter_rows(named=True):\n",
    "            sample_code = str(row['sample'])\n",
    "            project_id = str(row['project_id.project']).split('-')[1]\n",
    "            sample_type = str(row['tissue_type.samples'])\n",
    "            vital_status = int(str(row['vital_status.demographic']).lower() == 'dead')\n",
    "            days_to_event = row[survival_time_column[vital_status]]\n",
    "            days_to_event = int(days_to_event) if days_to_event is not None else 0\n",
    "            if sample_type == 'Normal':\n",
    "                project_id = 'NORMAL'\n",
    "\n",
    "            if sample_code not in tcga_barcodes:\n",
    "                continue\n",
    "            barcode_project_list.append((sample_code, project_id, vital_status, days_to_event))\n",
    "\n",
    "    barcode_project_df = pl.DataFrame(barcode_project_list, schema=[\"Barcode\", \"Tumor\", 'Vital_Status', 'Days_to_event'], orient='row')\n",
    "    barcode_project_df = barcode_project_df.with_columns(\n",
    "        pl.col('Tumor').cast(pl.Categorical).to_physical().alias('Tumor_codes')\n",
    "    )\n",
    "    \n",
    "    os.makedirs(os.path.dirname(output_file_path), exist_ok=True)\n",
    "    barcode_project_df.write_csv(output_file_path, separator='\\t')\n",
    "    return barcode_project_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_shape = create_clinical_data_df()\n",
    "print(df_shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python_3.13.3_Kernel",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
